### PR TITLE
Create Block deserializer & add deserializeBlock function

### PR DESCRIPTION
**Pull request by TrustHenry**
Deserialize a struct and return it as a Block.

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
a bad comment.

File: source/agora/common/Block.d [hunk 1]
```diff
@@ -74,6 +75,22 @@ public struct BlockHeader
...
+        Params:
+            dg = deserialize function accumulator
+
+    ***************************************************************************/
+
+    public void deserialize (scope DeserializeDg dg) nothrow @safe
+    {
+        this.prev_block = Hash(dg(64));
+        deserializePart(this.height, dg);
+        this.tx_hash = Hash(dg(64));
```
Comment (AndrejMitrovic) [inline source/agora/common/Block.d]: Maybe:

```suggestion
        this.tx_hash = Hash(dg(Hash.sizeof));
```

File: source/agora/common/Block.d [hunk 2]
```diff
@@ -74,6 +75,22 @@ public struct BlockHeader
...
+        Block Header Deserialization
+
+        Params:
+            dg = deserialize function accumulator
+
+    ***************************************************************************/
+
+    public void deserialize (scope DeserializeDg dg) nothrow @safe
+    {
+        this.prev_block = Hash(dg(64));
```
Comment (AndrejMitrovic) [inline source/agora/common/Block.d]:
```suggestion
        this.prev_block = Hash(dg(Hash.sizeof));
```

File: source/agora/common/Block.d [hunk 3]
```diff
@@ -130,10 +147,35 @@ public struct Block
...
+    public void deserialize (scope DeserializeDg dg) nothrow @safe
+    {
+        deserializePart(this.header, dg);
+        ulong size;
+        deserializePart(size, dg);
+        this.txs.length = size;
+        foreach (idx; 0 .. size)
+        {
+            Transaction tx;
+            deserializePart(tx, dg);
```
Comment (AndrejMitrovic) [inline source/agora/common/Block.d]:
```suggestion
            deserializePart(this.txs[idx], dg);
```

And remove the `Transaction tx;` ?

File: source/agora/common/Block.d [hunk 4]
```diff
@@ -130,10 +147,35 @@ public struct Block
...
+            this.txs[idx] = tx;
+        }
+        buildMerkleTree();
```
Comment (AndrejMitrovic) [inline source/agora/common/Block.d]:
```suggestion
        this.buildMerkleTree();
```

File: source/agora/common/Block.d (hunk 2 above)
Comment (TrustHenry) [reply source/agora/common/Block.d]: That's correct.

File: source/agora/common/Deserializer.d [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+public alias DeserializeDg = ubyte[] delegate(size_t size) nothrow @safe;
+
+/*******************************************************************************
+
+    Deserialize a struct and return it as a Block.
+
+    Params:
+        T = Type of struct to deserialize
+        record = Instance of `T` to deserialize
+        dg  = Deserialization delegate, when this struct is a nested struct
```
Comment (AndrejMitrovic) [inline source/agora/common/Deserializer.d:31]: `source` is undocumented

File: source/agora/common/Deserializer.d [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    Returns:
+        The deserialized `Block`
+
+*******************************************************************************/
+
+public Block deserializeBlock (scope const ref ubyte [] source)
+    /*pure*/ nothrow @trusted
+{
+    Block block;
+    ubyte[] data = source.dup();
```
Comment (AndrejMitrovic) [inline source/agora/common/Deserializer.d]: Why is there a need to duplicate the source?

File: source/agora/common/Deserializer.d [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+                "Struct `" ~ T.stringof ~
+                "` does not implement `deserialize(scope DeserializeDg) nothrow` function");
+    record.deserialize(dg);
+}
+
+/// Ditto
+public void deserializePart (ref Hash record, scope DeserializeDg dg)
+    nothrow @safe
+{
+    record = Hash(dg(64));
```
Comment (AndrejMitrovic) [inline source/agora/common/Deserializer.d]:
```suggestion
    record = Hash(dg(Hash.sizeof));
```

File: source/agora/common/Deserializer.d (hunk 5 above)
Comment (TrustHenry) [reply source/agora/common/Deserializer.d:31]: record -> source = uByte[] source to be deserialized

File: source/agora/common/Transaction.d [hunk 8]
```diff
@@ -90,12 +91,50 @@ public struct Transaction
 
     public void serialize (scope SerializeDg dg) pure const nothrow @safe
     {
+        serializePart(this.inputs.length, dg);
```
Comment (AndrejMitrovic) [inline source/agora/common/Transaction.d]: So this was missing in the serialization part? It should be in its own commit then.

Comment (AndrejMitrovic) [reply source/agora/common/Transaction.d]: Unless @Geod24 approves it.

File: source/agora/common/Transaction.d [hunk 9]
```diff
@@ -90,12 +91,50 @@ public struct Transaction
...
+        Transactions deserialization
+
+        Params:
+            dg = Deserialize function
+
+    ***************************************************************************/
+
+    public void deserialize (scope DeserializeDg dg) nothrow @safe
+    {
+        ulong inputSize;
```
Comment (AndrejMitrovic) [inline source/agora/common/Transaction.d]:
```suggestion
        ulong input_size;
```

Style guide (sorry)

File: source/agora/common/Transaction.d [hunk 10]
```diff
@@ -90,12 +91,50 @@ public struct Transaction
...
+        deserializePart(inputSize, dg);
+        this.inputs.length = inputSize;
+
+        // deserialize and generate inputs
+        foreach (idi; 0 .. inputSize)
```
Comment (AndrejMitrovic) [inline source/agora/common/Transaction.d]: Either `idx` or `input_idx`, I don't want us to have to guess what "idi" and "ido" mean.

File: source/agora/common/Transaction.d [hunk 11]
```diff
@@ -90,12 +91,50 @@ public struct Transaction
...
+            deserializePart(input , dg);
+            this.inputs[idi] = input;
+        }
+
+        ulong outputSize;
+        deserializePart(outputSize, dg);
+        this.outputs.length = outputSize;
+
+        // deserialize and generate outputs
+        foreach (ido; 0 .. outputSize)
```
Comment (AndrejMitrovic) [inline source/agora/common/Transaction.d]: Same comment, `idx` or `idx_output`

File: source/agora/common/Transaction.d [hunk 12]
```diff
@@ -202,6 +256,22 @@ public struct Input
...
+        Input deserialization
+
+        Params:
+             dg = deserialize function
+
+    ***************************************************************************/
+
+    public void deserialize (scope DeserializeDg dg) nothrow @safe
+    {
+        this.previous = Hash(dg(64));
```
Comment (AndrejMitrovic) [inline source/agora/common/Transaction.d]:
```suggestion
        this.previous = Hash(dg(Hash.sizeof));
```

File: source/agora/common/Transaction.d [hunk 13]
```diff
@@ -202,6 +256,22 @@ public struct Input
...
+        deserializePart(this.index, dg);
+        this.signature = Hash(dg(64));
```
Comment (AndrejMitrovic) [inline source/agora/common/Transaction.d]:
```suggestion
        this.signature = Hash(dg(Hash.sizeof));
```

File: source/agora/common/Deserializer.d (hunk 6 above)
Comment (TrustHenry) [reply source/agora/common/Deserializer.d]: I wanted to isolate the memory area.

Comment (TrustHenry) [reply source/agora/common/Deserializer.d]: Is it a waste of memory?

File: source/agora/common/Block.d (hunk 4 above)
Comment (TrustHenry) [reply source/agora/common/Block.d]: add on.

File: source/agora/common/crypto/Key.d [hunk 14]
```diff
@@ -190,6 +191,24 @@ public struct PublicKey
...
+
+        Returns:
+            `PublicKey` Public key address
+
+    ***************************************************************************/
+
+    public static PublicKey fromBinary (scope DeserializeDg dg) nothrow @safe
+    {
+        alias DType = typeof(this.data);
+        return PublicKey(DType(dg(DType.Width)));
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/Key.d]:
```suggestion
        return PublicKey(DType(dg(DType.sizeof)));
```

File: source/agora/node/Ledger.d [hunk 15]
```diff
@@ -232,21 +233,10 @@ unittest
...
-    ~"CDE27387366F9A64ABC91788162EFE5B3C9DAA4104113445994C7F0497199FF442E88F78C582A2CF"
-    ~"0000000066D002DC3D4F02352BF1E9478501CCF78FEA23AAFBAD4652630297C030A2F53E028BE2CD"
-    ~"CEEBADCCAD09908A375495919B49F263127B9BCC06DC8519A51EF80D005A6202000000009D0238E0"
-    ~"A171400BC6D68A9D9B316ACD5109649113A05C284F4296D2B30122F5";
-
-    assert(data.toHexString() == serializeData);
+    // getLastBlock Testing serialization & deserialization
+    // Compare the serialization data with the origin Ledger data.
+    const ubyte[] serializedData = serializeFull(ledger.getLastBlock());
+    assert(ledger.getLastBlock == deserializeBlock(serializedData));
```
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: Nice!

File: source/agora/common/Transaction.d (hunk 9 above)
Comment (TrustHenry) [reply source/agora/common/Transaction.d]: Yes,

File: source/agora/common/Transaction.d (hunk 10 above)
Comment (TrustHenry) [reply source/agora/common/Transaction.d]: modify idx

File: source/agora/common/Transaction.d (hunk 11 above)
Comment (TrustHenry) [reply source/agora/common/Transaction.d]: use same idx

File: source/agora/common/Transaction.d (hunk 12 above)
Comment (TrustHenry) [reply source/agora/common/Transaction.d]: That's it

File: source/agora/common/Transaction.d (hunk 13 above)
Comment (TrustHenry) [reply source/agora/common/Transaction.d]: Yes

File: source/agora/common/crypto/Key.d (hunk 14 above)
Comment (TrustHenry) [reply source/agora/common/crypto/Key.d]: @Geod24 That's what he told me.

Comment (AndrejMitrovic) [reply source/agora/common/crypto/Key.d]: Well it shouldn't really make a difference, except `.sizeof` can always be trusted. But here `.sizeof == .Width`, so use whichever one you prefer.

Comment (Geod24) [reply source/agora/common/crypto/Key.d]: `.sizeof` is better ;)

File: source/agora/common/Transaction.d (hunk 8 above)
Comment (TrustHenry) [reply source/agora/common/Transaction.d]: This is added this time.;;

File: source/agora/common/crypto/Key.d (hunk 14 above)
Comment (TrustHenry) [reply source/agora/common/crypto/Key.d]: I use .sizeof

File: source/agora/common/Block.d (hunk 1 above)
Comment (TrustHenry) [reply source/agora/common/Block.d]: sure

File: source/agora/node/Ledger.d (hunk 15 above)
Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d]: I would suggest moving this test somewhere else, like in `Serialize.d` or `Deserialize.d`. It's only serializing a block, it doesn't need the Ledger class for anything here. You can make your own block in a separate unittest.

Comment (TrustHenry) [reply source/agora/node/Ledger.d]: I will proceed with the test code on my own in the next PR  :)

File: source/agora/common/Block.d (hunk 3 above)
Comment (TrustHenry) [reply source/agora/common/Block.d]: use it,

File: source/agora/common/Deserializer.d (hunk 7 above)
Comment (TrustHenry) [reply source/agora/common/Deserializer.d]: changed

File: source/agora/common/Block.d [hunk 16]
```diff
@@ -130,10 +147,33 @@ public struct Block
...
+    {
+        deserializePart(this.header, dg);
+        ulong size;
+        deserializePart(size, dg);
+        this.txs.length = size;
+        foreach (idx; 0 .. size)
+        {
+            deserializePart(this.txs[idx], dg);
+        }
+        this.buildMerkleTree();
```
Comment (Geod24) [inline source/agora/common/Block.d]: This is going to be expensive on the long run

File: source/agora/common/Deserializer.d (hunk 6 above)
Comment (Geod24) [reply source/agora/common/Deserializer.d]: Yes. We don't need it to be `ref` either

File: source/agora/common/Deserializer.d [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+public Block deserializeBlock (scope const ref ubyte [] source)
+    /*pure*/ nothrow @trusted
+{
+    Block block;
+    ubyte[] data = source.dup();
+    size_t offset = 0;
+
+    DeserializeDg dg = (size) nothrow @safe
+    {
+        ubyte [] res = data[offset .. offset + size];
```
Comment (Geod24) [inline source/agora/common/Deserializer.d]:
```suggestion
        ubyte[] res = data[offset .. offset + size];
```

File: source/agora/common/Deserializer.d [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+*******************************************************************************/
+
+public Block deserializeBlock (scope const ref ubyte [] source)
+    /*pure*/ nothrow @trusted
+{
+    Block block;
+    ubyte[] data = source.dup();
+    size_t offset = 0;
+
+    DeserializeDg dg = (size) nothrow @safe
```
Comment (Geod24) [inline source/agora/common/Deserializer.d]:
```suggestion
    scope DeserializeDg dg = (size) nothrow @safe
```

**Review by Geod24 — COMMENTED** [review]
Separating into smaller commits would be great as @AndrejMitrovic mentioned.
At least the style fixes and the parts that fix serialization.

File: source/agora/common/Deserializer.d (hunk 17 above)
Comment (TrustHenry) [reply source/agora/common/Deserializer.d]: delete it

File: source/agora/common/Deserializer.d (hunk 18 above)
Comment (TrustHenry) [reply source/agora/common/Deserializer.d]: add on

File: source/agora/node/Ledger.d (hunk 15 above)
Comment (TrustHenry) [reply source/agora/node/Ledger.d]: Move to Deserializer

File: source/agora/common/Block.d (hunk 16 above)
Comment (TrustHenry) [reply source/agora/common/Block.d]: add on Serialize MekleTree[] & deserialize

File: source/agora/common/Deserializer.d (hunk 17 above)
Comment (TrustHenry) [reply source/agora/common/Deserializer.d]: ubyte[] data = source.dup(); -> Use ref

File: source/agora/common/Deserializer.d [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+        T = Type of struct to deserialize
+        source = uByte[] source to be deserialized
+        dg  = Deserialization delegate, when this struct is a nested struct
+
+    Returns:
+        The deserialized `Block`
+
+*******************************************************************************/
+
+public Block deserializeBlock (scope ref ubyte [] data)
```
Comment (Geod24) [inline source/agora/common/Deserializer.d]:
```suggestion
public Block deserializeBlock (scope ubyte[] data)
```

I don't know if we can make this data `const`. Probably not.

File: source/agora/common/Deserializer.d [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+    auto signature = genesis_key_pair.secret.sign(hashFull(tx)[]);
+    tx.inputs[0].signature = signature;
+
+    assert(ledger.acceptTransaction(tx));
+    ledger.makeBlock();
+    assert(ledger.getLastBlock().txs[$-1] == tx);
+
+    // getLastBlock Testing serialization & deserialization
+    // Compare the serialization data with the origin Ledger data.
+    ubyte[] serializedData = serializeFull(ledger.getLastBlock());
```
Comment (AndrejMitrovic) [inline source/agora/common/Deserializer.d]: Why do you need the `Ledger` if all you are doing is serializing a single block? And the `acceptTransaction` and unrelated stuff above is not relevant to serialization.

The serialization code is only concerned that it accurately serializes whatever data is in the Block to storage, and deserializes it properly. It doesn't care about validation of signatures, for example.

File: source/agora/common/Deserializer.d (hunk 19 above)
Comment (TrustHenry) [reply source/agora/common/Deserializer.d]: Applied the suggestion

File: source/agora/common/Deserializer.d (hunk 20 above)
Comment (TrustHenry) [reply source/agora/common/Deserializer.d]: Add simple unitTest to Deserializer

**Comment by TrustHenry** [comment]
Change the tx_hash -> merkle_root

File: source/agora/common/Block.d [hunk 21]
```diff
@@ -132,8 +149,39 @@ public struct Block
...
     {
         serializePart(this.header, dg);
+
+        serializePart(this.txs.length, dg);
         foreach (ref tx; this.txs)
             serializePart(tx, dg);
+
+        serializePart(this.merkle_tree.length, dg);
+        foreach (ref merkle; this.merkle_tree)
+            dg(merkle[]);
```
Comment (Geod24) [inline source/agora/common/Block.d:159]: @TrustHenry : This diff is fixing serialization, not deserialization. Please split it into atomic commits.
Let's do this together when you have a bit of time.

File: source/agora/common/Deserializer.d [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+/// Type of delegate deserializeDg
+public alias DeserializeDg = ubyte[] delegate(size_t size) nothrow @safe;
+
+/*******************************************************************************
+
+    Deserialize a struct and return it as a Block.
+
+    Params:
+        T = Type of struct to deserialize
+        source = uByte[] source to be deserialized
```
Comment (Geod24) [inline source/agora/common/Deserializer.d]:
```suggestion
        data = Binary serialized representation of `T` to be deserialized
```

File: source/agora/common/Deserializer.d [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+import agora.common.crypto.Key;
+
+/// Type of delegate deserializeDg
+public alias DeserializeDg = ubyte[] delegate(size_t size) nothrow @safe;
+
+/*******************************************************************************
+
+    Deserialize a struct and return it as a Block.
+
+    Params:
```
Comment (Geod24) [inline source/agora/common/Deserializer.d:27]: `record` is undocumented

File: source/agora/common/Deserializer.d [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+        source = uByte[] source to be deserialized
+        dg  = Deserialization delegate, when this struct is a nested struct
+
+    Returns:
+        The deserialized `Block`
+
+*******************************************************************************/
+
+public Block deserializeBlock (scope ubyte[] data)
+    /*pure*/ nothrow @trusted
```
Comment (Geod24) [inline source/agora/common/Deserializer.d]: Why `@trusted` ?

Comment (Geod24) [reply source/agora/common/Deserializer.d]: And why did you add `/*pure*/` ?
In the code you used, `/*pure*/` was added because the code *should* be pure, but cannot be made so (because some C function is called). This is not the case here.

File: source/agora/node/Ledger.d [hunk 25]
```diff
@@ -231,22 +229,6 @@ unittest
...
-     "02A91D85E7279998D31FE887281D0031D8C19FA732DC0BAC3A323569B19A14AD4B274255397A9A56"
-    ~"F14930CFAF7C7648377602EF0EF0E01821EE18F676092D060100000000000000DD9AAF1064DA6746"
-    ~"2DC1CC496E0374084058403429A1F74B72663F6CCA35B80AECD85063383D372F9DBCD770EF654B96"
-    ~"99F8C8801E6FD8041E9F517DF0B7CD2DE8D87AE11BDB8A861F7B0BD6EE6D7D37657FDD47CF8ED340"
-    ~"CDE27387366F9A64ABC91788162EFE5B3C9DAA4104113445994C7F0497199FF442E88F78C582A2CF"
-    ~"0000000066D002DC3D4F02352BF1E9478501CCF78FEA23AAFBAD4652630297C030A2F53E028BE2CD"
-    ~"CEEBADCCAD09908A375495919B49F263127B9BCC06DC8519A51EF80D005A6202000000009D0238E0"
-    ~"A171400BC6D68A9D9B316ACD5109649113A05C284F4296D2B30122F5";
-
-    assert(data.toHexString() == serializeData);
```
Comment (Geod24) [inline source/agora/node/Ledger.d:251]: Why is this removed ?

File: source/agora/common/Deserializer.d (hunk 24 above)
Comment (TrustHenry) [reply source/agora/common/Deserializer.d]: Fixed

File: source/agora/common/Deserializer.d (hunk 23 above)
Comment (TrustHenry) [reply source/agora/common/Deserializer.d:27]: add

File: source/agora/common/Deserializer.d (hunk 22 above)
Comment (TrustHenry) [reply source/agora/common/Deserializer.d]: Fixed

File: source/agora/node/Ledger.d (hunk 25 above)
Comment (TrustHenry) [reply source/agora/node/Ledger.d:251]: Move to Deserialzer.d But Not used to HexString
Compare the serialization data with the derialization data

File: source/agora/common/Transaction.d [hunk 26]
```diff
@@ -98,6 +99,42 @@ public struct Transaction
...
+        deserializePart(output_size, dg);
+        this.outputs.length = output_size;
+
+        // deserialize and generate outputs
+        foreach (idx; 0 .. output_size)
+        {
+            Output output;
+            deserializePart(output, dg);
+            this.outputs[idx] = output;
+        }
```
Comment (Geod24) [inline source/agora/common/Transaction.d:136]:
```d
// deserialize and generate outputs
foreach (ref output; this.outputs)
    deserializePart(output, dg);
```

File: source/agora/common/Transaction.d [hunk 27]
```diff
@@ -98,6 +99,42 @@ public struct Transaction
...
+        deserializePart(input_size, dg);
+        this.inputs.length = input_size;
+
+        // deserialize and generate inputs
+        foreach (idx; 0 .. input_size)
+        {
+            Input input;
+            deserializePart(input , dg);
+            this.inputs[idx] = input;
+        }
```
Comment (Geod24) [inline source/agora/common/Transaction.d:124]: See the way `output` was rewritten

File: source/agora/node/Ledger.d (hunk 25 above)
Comment (Geod24) [reply source/agora/node/Ledger.d:251]: The serialized data will be part of the interface in the future. But for now, let's go with it.

**Review by Geod24 — APPROVED** [review]

File: source/agora/common/Transaction.d (hunk 26 above)
Comment (TrustHenry) [reply source/agora/common/Transaction.d:136]: Applied the suggestion

File: source/agora/common/Transaction.d (hunk 27 above)
Comment (TrustHenry) [reply source/agora/common/Transaction.d:124]: Applied the suggestion
